nav-bar

nolixul opened the pull request:
Added a navbar component in a _components folder. The _(folder) prevents anything in that folder from being picked up by Next routing.
It's currently in the default color theme, but that'll change when I merge it in.
I'll also put hidden classes on the links for pages that aren't up yet when I merge.

<img width="1470" alt="Screenshot 2024-02-08 at 21 24 44" src="https://github.com/nolixul/postbox-library/assets/79650939/f7699e55-7a41-4476-84a7-17c3d9bee148">

nolixul committed:
Add font and basic navbar shell

nolixul committed:
Finished navbar

321jellyfish reviewed (changes requested):
Looks really pretty, I like the font and the colours. Please could you make the following changes? :)

Should: Add alt text to the image

Should: Use the Next Link component for the links and add the className to the Link component

Should: remove unnecessary comment lines 39 and 40

Could: rename class 'nav_links' to 'link_container' more semantic in my opinion

Should: change top <div> to be a <nav> element

Should: make logo and company names links back to the homepage

Should: needs to be made responsive - doesn't display well below 1005px screen

Question: current styling of nav links doesn't indicate they are links which isn't very accessible - could we underline? I know it looks less pretty but would be more accessible.

nolixul commented:
> Looks really pretty, I like the font and the colours. Please could you make the following changes? :)
> 
> Should: Add alt text to the image
> 
> Should: Use the Next Link component for the links and add the className to the Link component
> 
> Should: remove unnecessary comment lines 39 and 40
> 
> Could: rename class 'nav_links' to 'link_container' more semantic in my opinion
> 
> Should: change top
> 
> to be a element
> Should: make logo and company names links back to the homepage
> 
> Should: needs to be made responsive - doesn't display well below 1005px screen
> 
> Question: current styling of nav links doesn't indicate they are links which isn't very accessible - could we underline? I know it looks less pretty but would be more accessible.

Thanks! 
I left out the alt text because for logos you'd normally have the company name as the alt text, and the company name is right next to it. If we're going to have the logo as a link though, we'll have to put alt text in for sure. 
For me, the styling of the nav links where you get a mouse change, and colour change on hover makes it pretty clear you can click on them and something will happen. I take your point though - I've changed it from a list to a nav section, and made the header an actual <header> element so that everything is more semantically accurate. I've also added a bold on hover. If you still think an underline is necessary let me know :) 
Can you explain change top to be a element?
Everything else I agree with I'll go ahead and make those changes and then work on the merge conflicts, thanks for the feedback!

nolixul committed:
Made accessibility changes to navbar

nolixul committed:
Change out anchor for next links

nolixul committed:
remove comments

nolixul committed:
merge main

321jellyfish commented:
Sorry don't know where the text went for that comment. I thought you should add in a <nav> element instead of a <div> but I see you are doing that now. :)

I take your point about the nav and I think maybe we can not have the links underlined by default because it is a known pattern, but I do think we should have some indication that does not rely on colour to show that you are hovering over a link - could this be an underline? otherwise those with vision problems won't necessarily be able to tell that they are hovering over something and it is selected. 

Also I don't think you need to import react at the top of a component anymore.

nolixul committed:
Add navbar to layout

nolixul committed:
tidying up

nolixul committed:
Fix font import

nolixul commented:
All changes made :)

321jellyfish commented:
Thanks for making those changes.
Could you add the link to the Home page nav link? It's currently empty. 
Are you going to make the nav bar responsive in this PR?

nolixul commented:
> Thanks for making those changes. Could you add the link to the Home page nav link? It's currently empty. Are you going to make the nav bar responsive in this PR?

Yep, can do (although I will have to make an assumption about what all the page routes will be). I'll add bookshelf and add a book as well then :)

No, I've added another ticket for responsiveness, so I'll make a new PR for that!

nolixul committed:
Add routes

321jellyfish commented:
Just checked this out @nolixul and the links you have added are now making the styling on the nav links wrong for me. So it's displaying as blue (default link styling I assume.) Is this happening for you? 

Also is there a reason you've abbreviated the names of the pages in the url? I guess my opinion is it's nice to be consistent /addabook /ourbookshelf - what do you think is best for our users? 

I was wondering as well should the nav bar indicate to the user what page they are on? Could be a nice feature.

nolixul commented:
> Just checked this out @nolixul and the links you have added are now making the styling on the nav links wrong for me. So it's displaying as blue (default link styling I assume.) Is this happening for you?
> 
> Also is there a reason you've abbreviated the names of the pages in the url? I guess my opinion is it's nice to be consistent /addabook /ourbookshelf - what do you think is best for our users?
> 
> I was wondering as well should the nav bar indicate to the user what page they are on? Could be a nice feature.

- Yeah I see that styling too now, I'll get rid of that
- No, I just put in something off the top of my head, I'll change it to match what you said :)
- Good shout! I'll put it as the highlighted colour for now, and when styling firms up later we can smarten it up

321jellyfish commented:
@nolixul  I've been doing some research into typing React components please could you refactor so the return type is inferred? Following the example in the React docs. https://react.dev/learn/typescript also more info about React.FC type here: https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/function_components/ 

Unless there is a specific reason you want to type the function this way?

nolixul commented:
> @nolixul I've been doing some research into typing React components please could you refactor so the return type is inferred? Following the example in the React docs. https://react.dev/learn/typescript also more info about React.FC type here: https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/function_components/
> 
> Unless there is a specific reason you want to type the function this way?

No that's just an old habit from my last team, good to know how the docs show you're meant to do it - thanks :) will update

nolixul committed:
Merge branch 'main' into nav-bar

nolixul committed:
remove react component typing

nolixul committed:
Remove default link styles

nolixul committed:
changed link styles

321jellyfish commented on frontend/src/app/_components/navbar/Navbar.module.css:
```diff
@@ -0,0 +1,48 @@
+.link {
+    font-weight: 500;
+    text-decoration: none;
+    color: inherit;
```
Why have you added color: inherit?

nolixul replied:
This stops the links from being blue, it makes the text colour inherited from app instead. I thought better to set it as inherit so it can change based on global app styles than on the individual links :)

321jellyfish replied:
I don't think it will have that behaviour. "The inherit CSS keyword causes the element to take the computed value of the property from its parent element." [https://developer.mozilla.org/en-US/docs/Web/CSS/inherit] 

If you use inherit, it will inherit the color from it's parent element not globally.

But if you want to set it in global.css that is different and you are better off just doing that.

Play around with the code and you'll see what I mean i.e. set global link style to orange and keep the rest of your code the same.

321jellyfish replied:
Also using the global.css file keeps a clearer source of truth.

nolixul committed:
Change inherit to black

321jellyfish commented on frontend/src/app/_components/navbar/Navbar.module.css:
@nolixul you have only changed one of the color: inherit key/value paris - there are two more left in .active_link and .link:visited

nolixul committed:
align text

nolixul committed:
Change from inherit, add custom error page to remove dark mode
